haskell/src/Parser.hs

```diff
@@ -0,0 +1,43 @@
+module Parser (evaluateExpression) where
+
+import Text.ParserCombinators.Parsec
+import Nodes
+
+evaluateExpression  :: String -> Maybe ([Node])
+evaluateExpression input =
+    case parse parser "(unknown)" input of
+    Left parseError -> Nothing
+    Right result -> Just result
+
+parser :: Parser [Node]
+parser = expressionParser
+
+expressionParser :: Parser [Node]
+expressionParser = do
+    number <- numberParser
+    spaces
+    nodes <- many furtherExpressionParser
+    return $ [NumberNode (read number)] ++ (concat nodes)
+
+furtherExpressionParser :: Parser [Node]
+furtherExpressionParser = do
+    spaces
+    opChar <- operatorParser
+    spaces
+    number <- numberParser
+    return $ [OperatorNode (operatorFromChar opChar), NumberNode (read number)]
+
+integerParser :: Parser String
+integerParser = many digit
+
+numberParser :: Parser String
+numberParser = integerParser
```

Comment:
If you define this as `fmap read integerParser`, you'll get a `Parser Int` which is probably more useful. You could also use `fmap (NumberNode . read) integerParser` to get yourself a `Parser NumberNode`.

Parsers which return the type you want are usually easier to compose.